blocks/library/shortcode/index.js

```diff
@@ -2,14 +2,143 @@
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
-import { withInstanceId, Dashicon } from '@wordpress/components';
+import { withInstanceId, Dashicon, Spinner, SandBox } from '@wordpress/components';
+import { Component, compose } from '@wordpress/element';
+import { addQueryArgs } from '@wordpress/url';
+import BlockControls from '../../block-controls';
+import { getCurrentPostId } from '../../../editor/store/selectors';
+import { connect } from 'react-redux';
 
 /**
  * Internal dependencies
  */
 import './editor.scss';
 import PlainText from '../../plain-text';
 
+export class Shortcode extends Component {
+	constructor() {
+		super();
+		this.state = {
+			html: '',
+			js: '',
+			style: '',
+			type: '',
+			preview: false,
+			focus: false,
+			fetching: false,
+		};
+		this.doServerSideRender = this.doServerSideRender.bind( this );
+		this.handlePreviewClick = this.handlePreviewClick.bind( this );
+	}
+
+	doServerSideRender( event ) {
+		//This function sends the shortcode content and post ID to the rest endpoint,
+		//and retrieves the filtered shortcode content to be rendered in Preview
+
+		if ( event ) {
+			event.preventDefault();
+		}
+
+		//Get post ID from redux store
+		const postId = this.props.postId;
+		let shortcode = this.props.attributes.text;
+		shortcode = ( shortcode ) ? shortcode.trim() : '';
+		const apiUrl = addQueryArgs( wpApiSettings.root + 'gutenberg/v1/shortcodes', {
+			shortcode: shortcode,
+			postId: postId,
+			_wpnonce: wpApiSettings.nonce,
+
+		} );
+		if ( 0 === postId.length || null === postId ) {
+			//We don't have a post ID yet
+			this.setState( { html: __( 'Something went wrong. Try saving the post and try again' ) } );
+			return null;
+		}
+		if ( 0 === shortcode.length ) {
+			this.setState( { html: __( 'Enter something to preview' ) } );
+			return null;
+		}
+		this.setState( { fetching: true } );
+		window.fetch( apiUrl, {
+			credentials: 'include',
+		} ).then( ( response ) => {
+			response.json().then( ( { html, type, style, js } ) => {
+				html = ( 0 < html.length ) ? html : __( 'Sorry, couldn\'t render a preview' );
+				this.setState( { html: html, js: js, style: style, type: type, fetching: false } );
+			} );
+		} );
+	}
+
+	handlePreviewClick() {
+		this.setState( { preview: true } );
+		this.doServerSideRender();
+	}
+
+	render() {
+		const { fetching, preview, html, type, js, style } = this.state;
+		const { instanceId, setAttributes, attributes, focus, setFocus } = this.props;
+		const inputId = `blocks-shortcode-input-${ instanceId }`;
+
+		const controls = focus && (
```

Comment:
I think I've seen it mentioned elsewhere before, but in gutenberg isn't the default to show a preview, and only once the block is selected show the editor?

In that case I'm not sure these controls are helpful and if you unfocus from the block they don't appear on hover once the block is no longer selected

Reply:
@Lewiscowles1986 Please check the design discussion on the thread [here](https://github.com/WordPress/gutenberg/issues/1054#issuecomment-360271233). Please do share your thoughts on that thread. 
Also, these changes are outdated by newer revisions. Please check and review the most recent version of the code.